### PR TITLE
LP/NLP-based B&B algorithm with gurobi_persistent

### PR DESCRIPTION
**Pull request by ZedongPeng**
## Summary/Motivation:

MindtPy only supported LP/NLP-based B&B algorithm with cplex_persistent.
This PR adds the support of gurobi_persistent for (Regularized) LP/NLP-based B&B algorithm in MindtPy. 

There is one point that needs discussion in this PR. Since I need `solve_data` and `config` in the callback function, I defined a new SolverFactory `gurobi_persistent4mindtpy` and redefined `_intermediate_callback` function in m**ip_solve.py.** It works, but I am not sure if it is the best way. Any comments or suggestions are welcomed.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bernalde** [comment]
Hi, after the tests rerun all failures appear to be not our fault (unable to download BARON at some point). From yesterday's meeting, I heard that Jenkins was failing because of a missing import skip statement. Can you provide the log output of Jenkins to try to find it?

**Comment by jsiirola** [comment]
There are 8 failing jobs:
- pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_LPNLP
- pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_L1
- pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_L2
- pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_Linf
- pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_grad_lag
- pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_hess_lag
- pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_hess_only_lag
- pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_sqp_lag

They are all failing with errors similar to:
```
The pyomo.contrib.mindtpy.single_tree module (an optional Pyomo dependency) failed to import: ModuleNotFoundError: No module named 'cplex'
-------------------- >> begin captured logging << --------------------
pyomo.contrib.mindtpy: INFO: Single tree implementation is activated. The defalt MIP solver is cplex_persistent
pyomo.contrib.mindtpy: INFO: ---Starting MindtPy---
pyomo.contrib.mindtpy: INFO: Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
pyomo.contrib.mindtpy: INFO: rNLP is the initial strategy being used.

pyomo.contrib.mindtpy: INFO: Relaxed NLP: Solve relaxed integrality
pyomo.contrib.mindtpy: INFO: Relaxed NLP: OBJ: 15.082188768962084  LB: -inf  UB: inf  TIME:0.15s
pyomo.contrib.mindtpy: INFO: ---MindtPy main Iteration 1---
pyomo.contrib.mindtpy: INFO: MIP 1: Solve main problem.
--------------------- >> end captured logging << ---------------------
```

### Stack Trace

```
  File "[...]/python/3.8.7/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "[...]/python/3.8.7/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "[...]/python/3.8.7/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "[...]/pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py", line 57, in test_LPNLP
    results = opt.solve(model, strategy='OA',
  File "[...]/pyomo/contrib/mindtpy/MindtPy.py", line 171, in solve
    MindtPy_iteration_loop(solve_data, config)
  File "[...]/pyomo/contrib/mindtpy/iterate.py", line 57, in MindtPy_iteration_loop
    main_mip, main_mip_results = solve_main(solve_data, config)
  File "[...]/pyomo/contrib/mindtpy/mip_solve.py", line 68, in solve_main
    mainopt = setup_mip_solver(solve_data, config, regularization_problem)
  File "[...]/pyomo/contrib/mindtpy/mip_solve.py", line 165, in setup_mip_solver
    mainopt.set_callback(single_tree.LazyOACallback_gurobi)
  File "[...]/pyomo/common/dependencies.py", line 135, in __getattr__
    return getattr(_mod, attr)
  File "[...]/pyomo/common/dependencies.py", line 61, in __getattr__
    raise DeferredImportError(self._moduleunavailable_message())
The pyomo.contrib.mindtpy.single_tree module (an optional Pyomo dependency) failed to import: ModuleNotFoundError: No module named 'cplex'
```

**Comment by ZedongPeng** [comment]
> There are 8 failing jobs:
> 
> * pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_LPNLP
> * pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_L1
> * pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_L2
> * pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_Linf
> * pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_grad_lag
> * pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_hess_lag
> * pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_hess_only_lag
> * pyomo.contrib.mindtpy.tests.test_mindtpy_lp_nlp.TestMindtPy.test_RLPNLP_sqp_lag
> 
> They are all failing with errors similar to:
> 
> ```
> The pyomo.contrib.mindtpy.single_tree module (an optional Pyomo dependency) failed to import: ModuleNotFoundError: No module named 'cplex'
> -------------------- >> begin captured logging << --------------------
> pyomo.contrib.mindtpy: INFO: Single tree implementation is activated. The defalt MIP solver is cplex_persistent
> pyomo.contrib.mindtpy: INFO: ---Starting MindtPy---
> pyomo.contrib.mindtpy: INFO: Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
> pyomo.contrib.mindtpy: INFO: rNLP is the initial strategy being used.
> 
> pyomo.contrib.mindtpy: INFO: Relaxed NLP: Solve relaxed integrality
> pyomo.contrib.mindtpy: INFO: Relaxed NLP: OBJ: 15.082188768962084  LB: -inf  UB: inf  TIME:0.15s
> pyomo.contrib.mindtpy: INFO: ---MindtPy main Iteration 1---
> pyomo.contrib.mindtpy: INFO: MIP 1: Solve main problem.
> --------------------- >> end captured logging << ---------------------
> ```
> 
> ### Stack Trace
> ```
>   File "[...]/python/3.8.7/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
>     yield
>   File "[...]/python/3.8.7/lib/python3.8/unittest/case.py", line 676, in run
>     self._callTestMethod(testMethod)
>   File "[...]/python/3.8.7/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
>     method()
>   File "[...]/pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py", line 57, in test_LPNLP
>     results = opt.solve(model, strategy='OA',
>   File "[...]/pyomo/contrib/mindtpy/MindtPy.py", line 171, in solve
>     MindtPy_iteration_loop(solve_data, config)
>   File "[...]/pyomo/contrib/mindtpy/iterate.py", line 57, in MindtPy_iteration_loop
>     main_mip, main_mip_results = solve_main(solve_data, config)
>   File "[...]/pyomo/contrib/mindtpy/mip_solve.py", line 68, in solve_main
>     mainopt = setup_mip_solver(solve_data, config, regularization_problem)
>   File "[...]/pyomo/contrib/mindtpy/mip_solve.py", line 165, in setup_mip_solver
>     mainopt.set_callback(single_tree.LazyOACallback_gurobi)
>   File "[...]/pyomo/common/dependencies.py", line 135, in __getattr__
>     return getattr(_mod, attr)
>   File "[...]/pyomo/common/dependencies.py", line 61, in __getattr__
>     raise DeferredImportError(self._moduleunavailable_message())
> The pyomo.contrib.mindtpy.single_tree module (an optional Pyomo dependency) failed to import: ModuleNotFoundError: No module named 'cplex'
> ```

Hi John. Thanks for your response. All the errors in the Stack Trace are related to `import cplex`. Should I replace it with `attempt_import`. 

We have discussed this question before. https://github.com/Pyomo/pyomo/pull/1391#discussion_r415933190

**Comment by jsiirola** [comment]
@ZedongPeng yes: `attempt_import` can help, but is not a cure-all.  The fundamental issue is that Pyomo (intentionally) has a minimal list of required dependencies.  We require that the core functionality all work with the minimal dependencies.  So, optional packages (like those in contrib) can have required dependencies on additional packages (like cplex), as long as either the module is not imported by default when `pyomo.environ` is imported (pynumero does this), or the *use* of the optional dependency is guarded by importing the module through attempt_import and it is not triggered until the module actually tries to use the dependency (with tests appropriately guarded by `_available` checks).

File: pyomo/contrib/mindtpy/single_tree.py [hunk 1]
```diff
@@ -27,12 +27,11 @@
 from pyomo.opt import TerminationCondition as tc
 from pyomo.core import Constraint, minimize, value, maximize
 cplex, cplex_available = attempt_import('cplex')
-from cplex.callbacks import LazyConstraintCallback
 
 logger = logging.getLogger('pyomo.contrib.mindtpy')
 
 
-class LazyOACallback_cplex(LazyConstraintCallback):
+class LazyOACallback_cplex(cplex.callbacks.LazyConstraintCallback):
```
Comment (jsiirola) [inline pyomo/contrib/mindtpy/single_tree.py:34]: This won't help when cplex is not available, as `cplex.callbacks` will trigger the import of `cplex`.  Instead, try:
```python
class LazyOACallback_cplex(cplex.callbacks.LazyConstraintCallback if cplex_available else object):
```

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/single_tree.py:34]: Thanks

**Comment by bernalde** [comment]
I think the only failing test was because of a NEOS failure.

File: pyomo/contrib/mindtpy/mip_solve.py [hunk 2]
```diff
@@ -127,22 +130,39 @@ def setup_mip_solver(solve_data, config, regularization_problem):
     if regularization_problem:
         mainopt = SolverFactory(config.mip_regularization_solver)
     else:
-        mainopt = SolverFactory(config.mip_solver)
+        if config.mip_solver == 'gurobi_persistent' and config.single_tree:
+            @SolverFactory.register('gurobi_persistent4mindtpy', doc='Persistent python interface to Gurobi')
```
Comment (jsiirola) [inline pyomo/contrib/mindtpy/mip_solve.py:134]: Why are you registering this (temporary) solver with the SolverFactory? Do you ever see a case where a user would want to create another instance of this solver outside of mindpy?

I would just create an instance of this new solver and return it (and avoid the Factory)

File: pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py [hunk 3]
```diff
@@ -23,12 +23,18 @@
...
+required_nlp_solvers = 'ipopt'
+required_mip_solvers = ['cplex_persistent', 'gurobi_persistent']
+
+
+if SolverFactory(required_nlp_solvers).available(False) and any(SolverFactory(s).available(False) for s in required_mip_solvers):
     subsolvers_available = True
+    available_mip_solvers = [s for s in required_mip_solvers
+                             if SolverFactory(s).available(False)]
 else:
     subsolvers_available = False
```
Comment (jsiirola) [inline pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py:35]: My read of this is that the single tree tests use either cplex or gurobi - whichever happens to be present.  Can you consider having (a subset of) these tests duplicated and tied explicitly to Cplex and Gurobi?

**Review by jsiirola — COMMENTED** [review]
This looks pretty reasonable.  One thing that I think needs to be changed (the SolverFactory registration) and one that would be nice to have (separate tests explicitly tied to cplex_persistent and gurobi_persistent [for coverage])

File: pyomo/contrib/mindtpy/mip_solve.py (hunk 2 above)
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/mip_solve.py:134]: This solver will only be used in MindtPy. I thought this solver needs first to be registered for later use even just in MindtPy. Is there an example that only create a instance of the solver？

File: pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py (hunk 3 above)
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py:35]: No problem.

File: pyomo/contrib/mindtpy/mip_solve.py (hunk 2 above)
Comment (jsiirola) [reply pyomo/contrib/mindtpy/mip_solve.py:134]: All you need to do is make an instance of the class (that is all the SolverFactory does).  In this case, I think something like the following would work:
```python
mainopt = GurobiPersistent4MindtPy()
mainopt.solve_data = solve_data
mainopt.config = config
```
I also think that the solver can be declared at the module scope ... and doesn't have to be a "temporary" class declared in the method.

**Review by jsiirola — APPROVED** [review]

**Review by blnicho — APPROVED** [review]

**Pull request merged**